core/engine.go

```diff
@@ -106,128 +106,182 @@ func NewEngine(ex lib.ExecutionScheduler, o lib.Options, logger *logrus.Logger)
-	e.runLock.Lock()
-	defer e.runLock.Unlock()
+	// TODO: move all of this in a separate struct? see main TODO above
+
+	runSubCtx, runSubCancel := context.WithCancel(runCtx)
 
-	e.logger.Debug("Engine: Starting with parameters...")
+	resultCh := make(chan error)
+	runFn := func() error {
+		e.logger.Debug("Execution scheduler starting...")
+		err := e.ExecutionScheduler.Run(globalCtx, runSubCtx, e.Samples)
+		e.logger.WithError(err).Debug("Execution scheduler terminated")
 
-	collectorwg := sync.WaitGroup{}
-	collectorctx, collectorcancel := context.WithCancel(context.Background())
+		select {
+		case <-runSubCtx.Done():
+			// do nothing, the test run was aborted somehow
+		default:
+			resultCh <- err // we finished normally, so send the result
+		}
 
+		return err
+	}
+	waitFn := e.startBackgroundProcesses(globalCtx, runCtx, resultCh, runSubCancel)
+	return runFn, waitFn, nil
+}
+
+// This starts a bunch of goroutines to process metrics, thresholds, and set the
+// test run status when it ends. It returns a function that can be used after
+// the provided context is called, to wait for the complete winding down of all
+// started goroutines.
+func (e *Engine) startBackgroundProcesses( //nolint:funlen
+	globalCtx, runCtx context.Context, runResult <-chan error, runSubCancel func(),
+) (wait func()) {
+	processes := new(sync.WaitGroup)
+
+	// Spin up all configured collectors
 	for _, collector := range e.Collectors {
-		collectorwg.Add(1)
+		processes.Add(1)
 		go func(collector lib.Collector) {
-			collector.Run(collectorctx)
-			collectorwg.Done()
+			collector.Run(globalCtx)
+			processes.Done()
 		}(collector)
 	}
 
-	subctx, subcancel := context.WithCancel(context.Background())
-	subwg := sync.WaitGroup{}
+	// Siphon and handle all produced metric samples
+	processes.Add(1)
+	go func() {
+		defer processes.Done()
+		e.processMetrics(globalCtx)
+	}()
+
+	// Run VU metrics emission, only while the test is running.
+	// TODO: move? this seems like something the ExecutionScheduler should emit...
+	processes.Add(1)
+	go func() {
+		defer processes.Done()
+		e.logger.Debug("Starting emission of VU metrics...")
+		e.runMetricsEmission(runCtx)
+		e.logger.Debug("Metrics emission terminated")
+	}()
 
-	// Run metrics emission.
-	subwg.Add(1)
+	// Update the test run status when the test finishes
+	processes.Add(1)
+	thresholdAbortChan := make(chan struct{})
 	go func() {
-		e.runMetricsEmission(subctx)
-		e.logger.Debug("Engine: Emission terminated")
-		subwg.Done()
+		defer processes.Done()
+		select {
+		case err := <-runResult:
+			if err != nil {
+				e.logger.WithError(err).Debug("run: execution scheduler returned an error")
+				e.setRunStatus(lib.RunStatusAbortedSystem)
+			} else {
+				e.logger.Debug("run: execution scheduler terminated")
+				e.setRunStatus(lib.RunStatusFinished)
+			}
+		case <-runCtx.Done():
+			e.logger.Debug("run: context expired; exiting...")
+			e.setRunStatus(lib.RunStatusAbortedUser)
```

Comment:
Hhmm should `RunStatusAbortedUser` be set in this case? This could also have been a clean exit, no?

Reply:
if it was a clean exit, the `err := <-runResult` would be chosen, see https://github.com/loadimpact/k6/blob/85d8826a0a395bf1bb26808c5e187f46195c80bd/core/engine.go#L143